pysph/sph/bc/inlet_outlet_manager.py

```diff
@@ -0,0 +1,318 @@
+            performs the change in inlet particles to outlet
+            particles.
+
+        Parameters
+        -----------
+
+        particle_array : list
+            List of all particle_arrays
+        """
+        obj = []
+        from pysph.sph.bc.inlet import Inlet
+        from pysph.sph.bc.outlet import Outlet
+        for inlet in self.inletinfo:
+            i_name = inlet.pa_name
+            self.update_inlet_outlet_info(particle_array[i_name])
+            for fluid in self.fluids:
+                in1 = Inlet(
+                        particle_array[i_name],
+                        particle_array[fluid],
+                        inlet,
+                        self.kernel,
+                        self.dim,
+                        self.active_stages
+                        )
+            obj.append(in1)
+
+        for outlet in self.outletinfo:
+            o_name = outlet.pa_name
+            self.update_inlet_outlet_info(particle_array[o_name])
+            for fluid in self.fluids:
+                out1 = Outlet(
+                        particle_array[o_name],
+                        particle_array[fluid],
+                        outlet,
+                        self.kernel,
+                        self.dim,
+                        self.active_stages
+                        )
+            obj.append(out1)
+
+        return obj
+
+
+class IOEvaluate(Equation):
+    def __init__(self, dest, sources, x, y, z, xn, yn, zn,
+                 maxdist=1000.0):
+        """Compute ioid for the particles
+           0 : particle is in fluid
+           1 : particle is inside the inlet/outlet
+           2 : particle is out of inlet/outlet
+
+        parameters:
+        ----------
+
+        dest : str
+            destination particle array name
+        sources : list
+            List of source particle arrays
+        x : float
+            x coordinate of interface point
+        y : float
+            y coordinate of interface point
+        z : float
+            z coordinate of interface point
+        xn : float
+            x component of interface outward normal
+        yn : float
+            y component of interface outward normal
+        zn : float
+            z component of interface outward normal
+        maxdist : float
+            Maximum length of inlet/outlet
+
+        """
+        self.x = x
+        self.y = y
+        self.z = z
+        self.xn = xn
+        self.yn = yn
+        self.zn = zn
+        self.maxdist = maxdist
+
+        super(IOEvaluate, self).__init__(dest, sources)
+
+    def initialize(self, d_ioid, d_idx):
+        d_ioid[d_idx] = 1
+
+    def loop(self, d_idx, d_x, d_y, d_z, d_ioid, d_disp):
+        delx = d_x[d_idx] - self.x
+        dely = d_y[d_idx] - self.y
+        delz = d_z[d_idx] - self.z
+
+        d_disp[d_idx] = delx * self.xn + dely * self.yn + delz * self.zn
+
+        if ((d_disp[d_idx] > -0.000001) &
```

Comment:
This should really be `and` but it is OK.

Reply:
done